tests/signature.rs

```diff
@@ -1,41 +1,70 @@
-use fuel_crypto::{Error, Hasher, PublicKey, SecretKey, Signature};
+use fuel_crypto::{Error, Message, PublicKey, SecretKey, Signature};
 use rand::rngs::StdRng;
 use rand::SeedableRng;
 
 #[test]
-fn ecrecover() {
+fn recover() {
     let rng = &mut StdRng::seed_from_u64(8586);
 
     let message =
         b"A beast can never be as cruel as a human being, so artistically, so picturesquely cruel.";
 
-    let secret = SecretKey::random(rng);
-    let public = secret.public_key();
+    for _ in 0..100 {
+        let message = Message::new(message);
+
+        let secret = SecretKey::random(rng);
+        let public = secret.public_key();
+
+        let signature = Signature::sign(&secret, &message);
+        let recover = signature.recover(&message).expect("Failed to recover PK");
+
+        assert_eq!(public, recover);
+    }
+}
+
+#[test]
+fn verify() {
+    let rng = &mut StdRng::seed_from_u64(8586);
+
+    let message =
+        b"Music expresses that which cannot be put into words and that which cannot remain silent.";
+
+    for _ in 0..100 {
+        let message = Message::new(message);
+
+        let secret = SecretKey::random(rng);
+        let public = secret.public_key();
 
-    let signature = secret.sign(message);
-    let recover = PublicKey::recover(signature, message).expect("Failed to recover PK");
+        let signature = Signature::sign(&secret, &message);
 
-    assert_eq!(public, recover);
+        signature
+            .verify(&public, &message)
+            .expect("Failed to verify signature");
+    }
 }
 
```

Comment:
This seems quite costly to exhaustively fuzz the entire search space every test run. Using prop-tests we could randomly sample the distribution on each test run and easily track which inputs would cause a failure:

```suggestion

use proptest::prelude::{
    prop::test_runner::{RngAlgorithm, TestRng}, *,
};
prop_compose! {
    fn arb_rng()(
        bytes: [u8; 32],
    ) -> TestRng {
        TestRng::from_seed(RngAlgorithm::ChaCha, &bytes)
    }
}

proptest! {
    #![proptest_config(ProptestConfig::with_cases(50))]
    #[test]
    fn corrupted_signature(
        mut rng in arb_rng(),
        position in 0..Signature::LEN,
        bit in 0..7,
    ) {
        let message = b"When life itself seems lunatic, who knows where madness lies?";
        let message = Message::new(message);

        let secret = SecretKey::random(&mut rng);
        let public = secret.public_key();

        let signature = Signature::sign(&secret, &message);

        // Tamper a random bit of the signature
        //
        // The recover and verify operations should fail in all cases.

        let mut s = signature;
        let m = 1u8 << bit;
        s.as_mut()[position] ^= m;

        // check recover
        match s.recover(&message) {
            Ok(pk) => assert_ne!(public, pk),
            Err(Error::InvalidSignature) => (),
            Err(e) => panic!("Unexpected error: {}", e),
        }

        // check verify
        assert!(s.verify(&public, &message).is_err());
    }

    #[test]
    fn corrupted_public_key(
        mut rng in arb_rng(),
        position in 0..PublicKey::LEN,
        bit in 0..7,
    ) {
        let message = b"When life itself seems lunatic, who knows where madness lies?";
        let message = Message::new(message);

        let secret = SecretKey::random(&mut rng);
        let mut public = secret.public_key();

        let signature = Signature::sign(&secret, &message);

        // Tamper a random bit of the public key.
        //
        // The verify operations should fail in all cases.
        let m = 1u8 << bit;
        public.as_mut()[position] ^= m;
        assert!(signature.verify(&public, &message).is_err());
    }
}

```

Reply:
The goal of that test is to protect the signature from bit malleability, also considering that we perform custom encoding/decoding.

To cover a portion of the available space, it is indeed good to have some fuzzy tests. I added this issue here to implement it in a subsequent PR as improvement/enhancement: https://github.com/FuelLabs/fuel-crypto/issues/5